### PR TITLE
cuSpatial Python User Guide

### PR DESCRIPTION
**Pull request by thomcom**
This PR adds `docs/source/user_guide/index.ipynb`, a comprehensive description of the APIs available in cuspatial python and example code for each.

The APIs are separated in a table of contents by the logical structure we've settled on for their location in the API. The demonstrate uses the `cuspatial.` version of every API, none are accessed using their full python package address.

At the bottom I'm working on some examples of performing `contains` and `within`, DE-9IM operations that can be performed easily with `point-in-polygon`. Otherwise the document is complete.

Depends on https://github.com/rapidsai/cuspatial/pull/685 and https://github.com/rapidsai/cuspatial/pull/680

This contributes to or closes https://github.com/rapidsai/cuspatial/issues/599 by providing an example use case of all python APIs.

### CODE REVIEW

**Comment by thomcom** [comment]
This PR is nearly complete with an incoming `.ipynb`. I ran into a fairly serious error however that prevents slicing `GeoDataframes` that contain a `GeoSeries`, https://github.com/rapidsai/cuspatial/issues/676, which I'm working on in parallel before this PR can be completed.

File: docs/source/user_guide/users.ipynb [hunk 1]
```diff
@@ -0,0 +1,1750 @@
+{
```
Comment (isVoid) [inline docs/source/user_guide/users.ipynb:1]: <p>If it's for the purpose of demonstrating where to find the APIs, I think API reference is sufficient. Folder structure of cuspatial is inherently an implementation detail and is hard to track the changes if we document them here.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/cuspatial/pull/666/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (isVoid) [inline docs/source/user_guide/users.ipynb:1]: <p><code>A cuspatial GeoDataFrame is a collection of data types</code>? I think it should be a collection of cudf Series and GeoSeries. cuspatial doesn't have built-in <code>geometry</code> dtype at this point so perhaps leave it out.</p><p> </p><p>Also a general suggestion on adding links. We should use it sporadically and be consistent. For example <code>cudf.Series</code> should just use one link.</p><p> </p><blockquote>One of the most important features of cuspatial is that it is highly integrated with <code>cudf</code>.&nbsp;You can use any <code>cudf</code> operation on cuspatial metadata columns, and most row 1:1 operations will work with a <code>geometry</code> column.</blockquote><p> </p><p>I'm not sure I understand. What is <code>cuspatial metadata columns</code> ? What is <code>row 1:1 operations</code> ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/cuspatial/pull/666/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

**Comment by isVoid** [comment]
Why is ReviewNB only quoting the first line of the block??

**Comment by thomcom** [comment]
<p>Unfortunately I don't know exactly how easy it will be to use ReviewNB. I could revert to using a markdown file instead.</p>

**Comment by thomcom** [comment]
The last notebook committed appears to be bugged, I updated it and hopefully it is working better with `ReviewNB` now?

File: docs/source/user_guide/users.ipynb (hunk 1 above)
Comment (isVoid) [reply docs/source/user_guide/users.ipynb:1]: This comment corresponds to: `Geopandas and cudf integration` section.

**Comment by isVoid** [comment]
> Unfortunately I don't know exactly how easy it will be to use ReviewNB. I could revert to using a markdown file instead.

As long as you can figure out the part where the comment goes to then it should be fine with the review tools. It seems like you have figured out my comment regarding the folder structures. I also added the location for the second comment.

File: docs/source/user_guide/users.ipynb [hunk 2]
```diff
@@ -0,0 +1,1591 @@
...
```
Comment (isVoid) [inline docs/source/user_guide/users.ipynb:1]: <p>Just a proposal.. I think other notebooks tend to include an executable script to install required environment. Create a code cell and write:</p><pre class="ql-syntax" spellcheck="false">!conda create -n rapids-22.06 -c rapidsai -c nvidia -c conda-forge  \ 
    cuspatial=22.06 python=3.9 cudatoolkit=11.5 
</pre><p> </p><p>As for docker image, I'm not sure if we need to include it. Perhaps replace it with a line:</p><p><code>For other options to create RAPIDS environment, such as docker or build from source, see  <a href="..." target="_blank">RA</a>PIDS Release Selector</code></p><p> </p><p>How do we plan to track the version of these installation scripts? I've never tried to use <code>sed</code> to update a notebook before. Alternative is to include the conda script in a small breakout file <code>install_rapids_conda.sh</code> then do</p><pre class="ql-syntax" spellcheck="false">! chmod +x install_rapids_conda.sh &amp;&amp; ./install_rapids_conda.sh  
</pre><p>Then we can use https://github.com/rapidsai/cuspatial/blob/47ba833a139466028c8a756cc865b7c76d8ec586/ci/release/update-version.sh#L30-L31 to update <code>install_rapids_conda.sh</code>  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/cuspatial/pull/666/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (isVoid) [inline docs/source/user_guide/users.ipynb:1]: <blockquote>cuspatial uses <code>GeoArrow</code> buffers, a GPU-accelerated buffer type format that makes the   &nbsp;performance you get with cuspatial possible.</blockquote><p><code>GeoArrow</code> itself isn't GPU-accelerated, but is a format well suited for massively parallel programming model that works very well with GPU. </p><p> </p><blockquote>GeoArrow uses [PyArrow]( &nbsp;<a href="https://arrow.apache.org/docs/python/index.html" target="_blank">https://arrow.apache.org/d</a><a href=" &nbsp;https://arrow.apache.org/docs/python/index.html       " target="_blank">PyArrow</a></blockquote><p>GeoArrow extends <a href="link" target="_blank">Arrow</a> specification and introduces several new types suited for geometry application. Cuspatial implements GeoArrow by extending types in PyArrow and handles format conversion with existing PyArrow constructors.</p><p> </p><blockquote> Using an Arrow [DenseArray]( &nbsp;<a href="https://arrow.apache.org/docs/python/data.html#union-arrays" target="_blank">https://arrow.apache.org/docs/python/data.html#union-arrays</a>), &nbsp;GeoArrow stores heterogeneous   &nbsp;orderings of Features, similar to the [GeoSeries]( &nbsp;<a href="https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoSeries.html" target="_blank">https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoSeries.html</a>     ) used &nbsp;in [GeoPandas]( &nbsp;<a href="https://geopandas.org/en/stable/index.html" target="_blank">https://geopandas.org/en/stable/index.html</a>).</blockquote><p>What's similar? GeoPandas' geoseries also use <code>DenseArray</code> or <code>GeoPandas</code>  also supports heterogeneous orderings of features? Also, perhaps you mean heterogeneous type of column elements? "Heterogenous orderings of features" sounds like different order of features, which is supported even in usual series right?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/cuspatial/pull/666/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

**Comment by isVoid** [comment]
This looks all very promising and nice to read, I just left some comments mostly about the introduction section on the terms and phrases that I got confused about. The code section is mostly good.

File: docs/source/user_guide/users.ipynb [hunk 3]
```diff
@@ -0,0 +1,1591 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "a56dfe17-1739-4b20-85c9-fcb5902c1585",
+   "metadata": {},
+   "source": [
+    "### [cuspatial.polygon_bounding_boxes](https://docs.rapids.ai/api/cuspatial/nightly/api_docs/spatial.html#cuspatial.polygon_bounding_boxes)\n",
+    "\n",
+    "`polygon_bounding_boxes` supports more-complex geometry objects: `Polygon`s with multiple rings.  \n",
+    "The combination of `part_offset` and `ring_offset` allows us to determine where the exterior ring  \n",
+    "ends, which will be used for computing the bounding box."
```
Comment (harrism) [inline docs/source/user_guide/users.ipynb:555]:
```suggestion
    "`polygon_bounding_boxes` supports more complex geometry objects such as `Polygon`s with multiple rings.  \n",
    "The combination of `part_offset` and `ring_offset` allows the function to use only the exterior ring  \n",
    "for computing the bounding box."
```

Comment (harrism) [reply docs/source/user_guide/users.ipynb:555]: Although this implementation detail is perhaps not necessary in this guide.

File: docs/source/user_guide/users.ipynb [hunk 4]
```diff
@@ -0,0 +1,1591 @@
...
+   },
+   "cell_type": "markdown",
+   "id": "993d7566-203e-4b87-adad-088e2fd92eed",
+   "metadata": {},
+   "source": [
+    "### [cuspatial.haversine_distance](https://docs.rapids.ai/api/cuspatial/stable/api_docs/gis.html#cuspatial.haversine_distance)\n",
+    "\n",
+    "Spatial developers will commonly be familiar with haversine distance, the great circle distance  \n",
+    "between longitude and latitude pairs. cuspatial uses the `lon/lat` ordering to better reflect the  \n",
+    "cartesian coordinates of great circle coordinates: `x/y`.\n",
```
Comment (harrism) [inline docs/source/user_guide/users.ipynb:764]: They might not be familiar...

```suggestion
    "Haversine distance is the great circle distance  \n",
    "between longitude and latitude pairs. cuspatial uses the `lon/lat` ordering to better reflect the  \n",
    "cartesian coordinates of great circle coordinates: `x/y`.\n",
```

File: docs/source/user_guide/users.ipynb [hunk 5]
```diff
@@ -0,0 +1,1591 @@
...
...zVPIPtslu7YgHfu9q8bwHtaoIAAQIENoiAK4Ab5EQPdJiHjm73tkkXXx1N6HxLz8Z1Z0d/du9YN++qrjarILYQIECAAAECBAYrMI5wq/nt6nbn+OfqNYxwWwRWvXk87uv49+cW0XDTxpc6jnNIs42vBAgQIECAAIHBCFSE281NgTOOcNux56OowrWdCPryNejzLxufShvpu80aMGiSAAECBAgQGLrAQUkubAqbuop2QZIDBjS4elZxfPWvfteE1ItM6ai3fdsro6LgBvQPRFcJECBAgACBpA8Rbos8D59pCsAqAt+1wAO8u6P9Ty6wfU0RIECAAAECBNZU4OQkNzQFzTjCbec1PfLaNX5SM54qACupZFFL11XSmhvRQoAAAQIECBDotUBFpm3uKJQuTjL0TNuaWqni3yZvAz8yiqab96Qc2/GM4fZIPJl3HPYnQIAAAQIENpBATTt0dpL7mwLpr0nq5Y/JCZ6HzFITU08WgPX5yiQVYbfapezalz+q3dNX26D9CBAgQIAAAQJrLVC3KdsC5uEkm5LssdYHX+f2n5jkmo4isOLqVrvUvm1RWW8Ym8tztaL2I0CAAAECBNZMoCLczuu4dXlpkiPX7Kjbv+FKKGnf1q0CruLstiWBp7b9VEfxV89KHrb9h6kHBAgQIECAAIFHBapwOSvJHU3xck+SM7axCHq01WF9+lAz9vEVvCp+j5hhKPU85M862qi5Bt83w/42IUCAAAECBAism0BdmfpxR+FSVwL3Wbde9ONAH+9wGBeCdQv3Y0nenuS4JPWSx9tGz0lWkTjebvJ3FX+VgWwhQIAAAQIECPRCYFqE22+TnNCLHm6fTtTVukozmSzkVvP57iRv3T5DcFQCBAgQIECAwJYCXRFu9bZv3QYWU5a8cDQf4GoKv7rq97Uk+2/Jbg0BAgQIECBAYP0FXpTkoo6rW0OLcFsvubrN+8Ukt3WYTSsOv7JenXMcAgQIENgYAssy79rGOFv9GmXl23509LzaZNbtTUnq5Yea6NmydYEqng9OUhNj75qkpo+p/5MfTrJ7s+sHknyhWecrAQIECBAgQGDdBN6Y5HfNFayhR7itG94MB3pNkgcb35oz8ZQZ9rUJAQIECBAgQGChAtMi3L6b/z/jttCDbfDG3pGkIuQmbws/kOSYDe5i+AQIECBAgMA6CWyUCLd14pz5MB9sCs
```
Comment (harrism) [inline docs/source/user_guide/users.ipynb:821]:
```suggestion
    "\n",
    "`pairwise_linestring_distance` computes the distance between corresponding linestrings in two `GeoSeries` of linestrings of  \n",
    "length `n`. It returns the  \n",
    "minimum distance from any point in the first linestring of a pair to the nearest segment or point within the second linestring of the pair.\n",
    "\n",
```

Comment (harrism) [reply docs/source/user_guide/users.ipynb:821]: I don't know why the suggestion is appearing like this...

Comment (harrism) [inline docs/source/user_guide/users.ipynb:826]:
```suggestion
    "input, the first argument must be a monotonically increasing column of values from $0$ to  \n",
```

Comment (harrism) [reply docs/source/user_guide/users.ipynb:826]: Frustrating. This was a one-word change suggestion, showing up as a diff between an image and a sentence.

Comment (harrism) [inline docs/source/user_guide/users.ipynb:829]: polygons should use a different API (that we don't have yet), shouldn't they?  I think we should demonstrate the designed purpose of each API first.

Comment (harrism) [inline docs/source/user_guide/users.ipynb:900]:
```suggestion
    "example the minimum distance from a country's center to it's border is computed.\n",
```

Comment (harrism) [reply docs/source/user_guide/users.ipynb:900]: Oh GitHub....

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1049]:
```suggestion
    "origin defined by `{x_min, y_min}`. This can increase the likelihood of generating  \n",
```

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1050]:
```suggestion
    "well-separated quads.\n",
```

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1122]:
```suggestion
    "The results `point_indices` and `quadtree` are used by `point_in_polygon_quadtree` for  \n",
```

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1123]: Very confusing, and probable unnecessary.

```suggestion
```

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1192]:
```suggestion
    "original MultiPolygon row. This is left as an exercise.\n",
```

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1277]: I think the example is above, not below. Probably should reference via a link, not by vertical direction, in case things get reordered.

```suggestion
    "See the example above in *Indexing* to use `quadtree_point_in_polygon`.\n",
```

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1273]:
```suggestion
    "not yet map to GeoPandas, but with knowledge of cuSpatial data formats you can call  \n",
```

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1275]: It''s `quadtree_point_in_polygon` (the quadtree is on the points, not the polygons)

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1331]: `quadtree_point_in_polygon`. Also, this is already covered above, so link to it.

**Review by harrism — CHANGES_REQUESTED** [review]
This is a great. I have a number of edits / suggestions. GitHub started doing weird things with my suggestions towards the end (treating the original text as an encoded image).

My biggest concern is that this document is already getting quite long, and as cuSpatial's API grows it is going to get unwieldy. I think going forward we should explore how we can modularize it, and where possible, associate the interactive examples with the individual API doc sections/pages.

File: docs/source/user_guide/users.ipynb (hunk 1 above)
Comment (thomcom) [reply docs/source/user_guide/users.ipynb:1]: <p>Agreed and removed it, replaced it with a TOC</p>

File: docs/source/user_guide/users.ipynb (hunk 2 above)
Comment (thomcom) [reply docs/source/user_guide/users.ipynb:1]: Thanks for the comments, I've revised the paragraph according to your suggestions. WRT your last paragraph, I think that "Feature" refers specifically to the spatial concept of a geometry feature and it's metadata. I've revised that section as well, is it more clear?

File: docs/source/developer_guide/internals.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+cuspatial uses an implementation of the [GeoArrow Format](
+https://github.com/geopandas/geo-arrow-spec              ) for high-  
+performance calculations on the GPU. In general, features are collected  
+into homogeneous buffers based on feature type. GeoArrow specifies a data  
+structure for Points, MultiPoints, LineStrings, and Polygons. Algorithms
+such as spatial join via point-in-polygon take as arguments a buffer of
+points and a buffer of polygons.
+
+GeoArrow uses offset buffers to encode where each feature begins and ends.  
+Our algorithms will usually require that these offset buffers be passed
```
[the remark this answers is not in the document]
Comment (thomcom) [reply docs/source/developer_guide/internals.md]: This file is out of date and has been dropped.

File: docs/source/user_guide/users.ipynb [hunk 7]
```diff
@@ -0,0 +1,1591 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "005074d1-8d89-4bf7-8082-7ac961a82eff",
+   "metadata": {},
+   "source": [
+    "## Installing cuspatial\n",
+    "\n",
+    "Read the [RAPIDS Quickstart Guide](\n",
+    "https://rapids.ai/start.html     ) to learn how to install all RAPIDS libraries, including cuspatial.  \n",
```
[the remark this answers is not in the document]
Comment (thomcom) [reply docs/source/user_guide/users.ipynb:47]: `library_design.md`, `contributing_guide.md`, and `development_environment.md` in `docs/source/developer_guide` all use `cuspatial` as well as `source/developer_guide/contributing_guide.md`. I'd rather leave it as cuspatial most of the time if not always, what do you say? I find cuSpatial to be unwieldy, and cuspatial to be smooth.

File: docs/source/user_guide/users.ipynb [hunk 8]
```diff
@@ -0,0 +1,1591 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4b1251d1-558a-4899-8e7a-8066db0ad091",
+   "metadata": {},
+   "source": [
+    "## I/O\n",
+    "\n",
+    "cuspatial supports two modes of loading Feature coordinates. [cuspatial.read_polygon_shapefile](\n",
```
[the remark this answers is not in the document]
Comment (thomcom) [reply docs/source/user_guide/users.ipynb:102]: :+1:

[the remark this answers is not in the document]
Comment (thomcom) [reply docs/source/user_guide/users.ipynb:102]: I added

One can also create feature geometries directly using any Python buffer that supports
__array_interface__ for coordinates and their feature offsets.

File: docs/source/user_guide/users.ipynb [hunk 9]
```diff
@@ -0,0 +1,1591 @@
...
+    "https://docs.rapids.ai/api/cuspatial/stable/api_docs/geopandas_compatibility.html#cuspatial.GeoDataFrame               ) is a collection of [cudf](\n",
+    "https://docs.rapids.ai/api/cudf/stable/              ) [Series](\n",
+    "https://docs.rapids.ai/api/cudf/stable/api_docs/series.html    ) data types and\n",
+    "[cuspatial.GeoSeries](\n",
+    "https://docs.rapids.ai/api/cuspatial/stable/api_docs/geopandas_compatibility.html#cuspatial.GeoSeries             ) `\"geometry\"` dtypes. Both columns are stored on the GPU, with\n",
+    "`GeoSeries` being represented under the hood with `GeoArrow`.\n",
+    "\n",
+    "One of the most important features of cuspatial is that it is highly integrated with `cudf`.  \n",
+    "You can use any `cudf` operation on cuspatial non-feature columns, and most non-reduction  \n",
+    "operations will work with a `geometry` column."
```
[the remark this answers is not in the document]
Comment (thomcom) [reply docs/source/user_guide/users.ipynb:278]: I added

Operations that reduce or collate the number of rows in your DataFrame, for example `groupby`, are not supported at this time.

File: docs/source/user_guide/users.ipynb [hunk 10]
```diff
@@ -0,0 +1,1591 @@
...
+   "source": [
+    "## Bounding\n",
+    "\n",
+    "Compute the bounding boxes of `n` polygons or linestrings:\n",
+    "\n",
+    "<img src=\"attachment:8c5d8b90-2241-45c2-b98c-20d48d1ee6b7.png\" width=250>\n",
+    "\n",
+    "### [cuspatial.trajectory_bounding_boxes](https://docs.rapids.ai/api/cuspatial/stable/api_docs/trajectory.html#cuspatial.trajectory_bounding_boxes)\n",
+    "\n",
+    "`trajectory_bounding_boxes` works `LineString`s and `Polygon`s with a single exterior ring.  \n",
```
[the remark this answers is not in the document]
Comment (thomcom) [reply docs/source/user_guide/users.ipynb:513]: You are right!

File: docs/source/user_guide/users.ipynb [hunk 11]
```diff
@@ -0,0 +1,1591 @@
...
+      "0  0.000043  0.000142  1.000046  0.999356\n",
+      "1  0.000065 -0.000022  0.999547  0.999639\n",
+      "2  0.000718 -0.000039  0.998866  0.999888\n",
+      "3  0.000212  0.000148  0.999976  1.000086\n",
+      "4 -0.000081  0.000646  0.998771  0.999750\n"
+     ]
+    }
+   ],
+   "source": [
+    "trajectory_bounding_boxes = cuspatial.core.spatial.bounding.polyline_bounding_boxes(\n",
```
[the remark this answers is not in the document]
Comment (thomcom) [reply docs/source/user_guide/users.ipynb:622]: I've had the same thought.

File: docs/source/user_guide/users.ipynb (hunk 5 above)
Comment (thomcom) [reply docs/source/user_guide/users.ipynb]: I will endeavor to process your comments in these sections that have diffed against the embedded images instead of the code.

Comment (thomcom) [reply docs/source/user_guide/users.ipynb:1123]: I think you're suggesting that I delete one of the images, tell me which one?

Comment (thomcom) [reply docs/source/user_guide/users.ipynb:1192]: There doesn't seem to be a suggestion to change here, let me know?

Comment (thomcom) [reply docs/source/user_guide/users.ipynb:1275]: Thanks!

Comment (thomcom) [reply docs/source/user_guide/users.ipynb]: Will do.

Comment (thomcom) [reply docs/source/user_guide/users.ipynb]: Not sure where this is directed.

File: docs/source/user_guide/users.ipynb (hunk 2 above)
Comment (thomcom) [reply docs/source/user_guide/users.ipynb:1]: Before we add a file as a dependency for the user guide, do you mind if we try to use the script or a small modification of it to update the `.ipynb` directly? It should work as it is just another text file.

File: docs/source/user_guide/users.ipynb [hunk 12]
```diff
@@ -0,0 +1,1591 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "caadf3ca-be3c-4523-877c-4c35dd25093a",
+   "metadata": {},
+   "source": [
+    "## Contents\n",
+    "\n",
+    "This guide provides a use example for all of the python API components of cuspatial.  \n",
```
[the remark this answers is not in the document]
Comment (thomcom) [reply docs/source/user_guide/users.ipynb:23]: I've modified `Installing cuspatial` to say

If you are working on a system with a CUDA-enabled GPU and have CUDA installed, uncomment the following cell and install cuspatial:

Do you think I should add more?

**Comment by thomcom** [comment]
I've addressed the existing reviews.

File: docs/source/user_guide/users.ipynb [hunk 13]
```diff
@@ -0,0 +1,1602 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7b770cb4-793e-467a-a306-2d3409545748",
+   "metadata": {},
+   "source": [
+    "## GPU accelerated memory layout\n",
+    "\n",
+    "cuspatial uses `GeoArrow` buffers, a GPU-friendly data format for geometric data that is well  \n",
+    "suited for massively parallel programming. See [I/O](#io) on the fastest methods to get your  \n",
```
Comment (harrism) [inline docs/source/user_guide/users.ipynb:81]: Test that this link works.

```suggestion
    "suited for massively parallel programming. See [I/O](#input-output) on the fastest methods to get your  \n",
```

File: docs/source/user_guide/users.ipynb (hunk 5 above)
Comment (harrism) [reply docs/source/user_guide/users.ipynb:1134]: No, I was deleting a sentence. Github screwed it up -- it keeps putting an encoded image as the "before" of the diff.

Comment (harrism) [reply docs/source/user_guide/users.ipynb:1134]: Let me try again.

File: docs/source/user_guide/users.ipynb [hunk 14]
```diff
@@ -0,0 +1,1602 @@
...
...0a62bDMRqGbAgoooIACCiigQBHYHjhwwFi8nGF7hzkpZWi46hJt9f9LM8xrUCzi28wwfZNSQAEFFFBAAQU6K5CzeI9uNMYybFsuoTLOJIxxAd7YyDMbgbNcu+/hA9J/9biFdH8FFFBAAQUUUKBPAldZI4TboiJpNMfp5UzgbWcInDGD/9JoBP5ohumblAIKKKCAAgoo0BmBtUK4HQnca0G1yMZZjverLv3m/5ypO+utuTZgRivZetaZmJ4CCiiggAIKKNBmgYyi0QzhluvkZUzfaUK4jVvnvRqNv2wA5hp+s97ePSCf6886E9NTQAEFFOivwMX6WzVrtgICVyiXez83ZxFlAAAN+klEQVQH7PH/27uzGFmqMg7gf1RUFmURRRRQQUUQFGR3RUFxARMXYmLQxMRgiFF8Q18ML0ZeNLkPGpdoDMYXfLzBFVzBhUUUlVUlqAiC7KAooObT7tCcW32n73TP3KqeXyWT6a6pOnXOr+5NvtRy/hPjrQi3o0cF4AMT69f6434dB7ixY928q+ot43YpCwsBAgQIEJhJQAE4E5ONeiYwjnCryZZPmejb7aMs3Xor9qqJ9ev1cc+OA1XM3KKXmruwXfZqV/hOgAABAgSmCUgRmCZjfV8Fjk/y2SSHT3Swnrurt2/rlu9dE+vX++NOHQfsKtY6NtumVV1tdh17mxq1MQECBAhsHAEF4MY510MfaV3h+vRoSpXJOLXLkpyZ5IqeDrCeA1z00tVmFcEWAgQIECAwk4BbwDMx2Wg7Cowj3K5L8p6JLN17k3wkyct7VPzVPIPtslu7YgHfu9q8bwHtaoIAAQIENoiAK4Ab5EQPdJiHjm73tkkXXx1N6HxLz8Z1Z0d/du9YN++qrjarILYQIECAAAECBAYrMI5wq/nt6nbn+OfqNYxwWwRWvXk87uv49+cW0XDTxpc6jnNIs42vBAgQIECAAIHBCFSE281NgTOOcNux56OowrWdCPryNejzLxufShvpu80aMGiSAAECBAgQGLrAQUkubAqbuop2QZIDBjS4elZxfPWvfteE1ItM6ai3fdsro6LgBvQPRFcJECBAgACBpA8Rbos8D59pCsAqAt+1wAO8u6P9Ty6wfU0RIECAAAECBNZU4OQkNzQFzTjCbec1PfLaNX5SM54qACupZFFL11XSmhvRQoAAAQIECBDotUBFpm3uKJQuT
```
Comment (harrism) [inline docs/source/user_guide/users.ipynb:1134]: This sentence is very confusing and doesn't add much. Suggest deleting (but when I make a suggestion github does the image diff thing).

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1135]: the function is called `quadtree_point_in_polygon`. Please search and replace all of `point_in_polygon_quadtree` with that.

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1284]: "The API surface for spatial joins does  \n",
    "not yet map to GeoPandas,"
    
-->
    
"The API does not yet match GeoPandas,"

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1288]: There's an empty set of single quotes here?

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1340]: cuSpatial (capital S).  As I commented before, this sentence seems redundant since you just discussed that other join algorithm above. At least link to that section.  Also, here's another `point_in_polygon_quadtree` that should be `quadtree_point_in_polygon`.d

Comment (harrism) [inline docs/source/user_guide/users.ipynb:1360]: It's weird that you keep saying this. This is the third time...

Comment (harrism) [reply docs/source/user_guide/users.ipynb:1360]: I think all the unfinished stuff below should be dropped from this PR.

**Review by harrism — CHANGES_REQUESTED** [review]

File: docs/source/user_guide/users.ipynb (hunk 14 above)
Comment (harrism) [reply docs/source/user_guide/users.ipynb:1134]: Ugh, so basically reviews are broken on this PR -- it associates every comment with line 819-820 no matter where I put them.

File: docs/source/user_guide/users.ipynb (hunk 2 above)
Comment (isVoid) [reply docs/source/user_guide/users.ipynb:1]: <p>Hmm, I still think the use of <code>ordering</code> seems a bit strange. I believe you mean union column supports "heterogeneous type of features". If so I think just say <code>type</code> is better.</p><p> </p><blockquote> be loaded and transformed in a method similar to the geoseries...</blockquote><p> </p><p>be loaded and transformed with methods similar to those in <code>GeoPandas.GeoSeries</code>  </p>

File: docs/source/user_guide/users.ipynb (hunk 1 above)
Comment (isVoid) [reply docs/source/user_guide/users.ipynb:1]: <p>Thanks. This looks much better now.</p>

File: docs/source/user_guide/users.ipynb [hunk 15]
```diff
@@ -0,0 +1,1602 @@
+{
```
Comment (isVoid) [inline docs/source/user_guide/users.ipynb:1]: <blockquote>a <code>Tuple</code> of GeoArrow buffers</blockquote><p> </p><p>But... they are just plain cudf series and dataframes!</p><p> </p><p>I would propose we implement <a href="https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoSeries.from_file.html" target="_blank">`GeoSeries.from_file</a>with this function.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/cuspatial/pull/666/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (isVoid) [inline docs/source/user_guide/users.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.    num_rings = len(geometry.polygons.ring_offset)
</pre><p>Some variables are unused in this cell <code>num_rings</code>,<code>points_x</code>,<code>points_y</code>,<code>points_offsets</code>        </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/rapidsai/cuspatial/pull/666/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

**Comment by harrism** [comment]
@thomcom I propose removing from this PR the speculative sections at the end on building up the rest of DE9IM (contains, etc) using existing primitives. That could be saved for a followup PR.

**Comment by thomcom** [comment]
I'll remove the stuff at the bottom and then merge.

File: docs/source/user_guide/users.ipynb (hunk 2 above)
Comment (isVoid) [reply docs/source/user_guide/users.ipynb:1]: Yes this is fine.

**Review by harrism — APPROVED** [review]

**Review by isVoid — CHANGES_REQUESTED** [review]
I have these comments still open:
https://github.com/rapidsai/cuspatial/pull/666#discussion_r981983039
https://github.com/rapidsai/cuspatial/pull/666#discussion_r982004977
https://github.com/rapidsai/cuspatial/pull/666#discussion_r982753469

File: docs/source/user_guide/users.ipynb (hunk 2 above)
Comment (thomcom) [reply docs/source/user_guide/users.ipynb:1]: Done.

File: docs/source/user_guide/users.ipynb (hunk 15 above)
Comment (thomcom) [reply docs/source/user_guide/users.ipynb:1]: I think I'd like to update `read_polygon_shapefile` to return a `GeoSeries`, as `from_file` implies many more file types to support. We could write `from_file` and check if it is a shapefile, then use our internal reader otherwise use `GeoPandas`, but performance would be poor for all other file types. What do you think? https://github.com/rapidsai/cuspatial/issues/715

Comment (thomcom) [reply docs/source/user_guide/users.ipynb:1]: Thanks for catching this.

**Review by isVoid — APPROVED** [review]

**Comment by thomcom** [comment]
@gpucibot merge
